lightning/src/ln/channel.rs

```diff
@@ -1539,38 +1512,39 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
 			panic!("Should not have advanced channel commitment tx numbers prior to funding_created");
 		}
 
-		let funding_txo = OutPoint{ txid: msg.funding_txid, index: msg.funding_output_index };
-		self.funding_txo = Some(funding_txo.clone());
+		let funding_txo = OutPoint { txid: msg.funding_txid, index: msg.funding_output_index };
+		self.channel_transaction_parameters.funding_outpoint = Some(funding_txo);
+		self.holder_keys.ready_channel(&self.channel_transaction_parameters);
```

Comment:
Right, so I'm not really sure what to do here - previously (and going all the way back to the first commit) we've been really careful to never update state in channel until we've checked everything we need to check. This resulted in the really awkward dance here where we assign the funding_txo, build a commitment, check the signature, and *then* maybe unset the funding_txo if the signature was bad. Its not actually entirely clear to me if this behavior (or the equivalent behavior in `get_outbound_funding_created`) ever mattered, or was externally-observable. With this line, its somewhat confused - we try very hard to make sure there are no irreversible actions taken before we check the message, but call `ready_channel` which is *very* externally observable.

These days its pretty clear that it shouldn't matter, and probably we should just add a comment noting we change something before checking, but then assert that if `funding_created_signature` (or `get_outbound_funding_created_signature`) return an Err it is a `ChannelError::Close`, implying there's no risk of accidentally using a partial state later, as the channel is closing.

Reply:
- would making `ready_channel` idempotent be helpful in any way?
- I don't quite get the solution in the last paragraph.  A `ChannelError::Close` return value will cause the caller to go into the closing flow.  So we haven't actually prevented a close from happening with a signer that got a `ready_channel`, right?
- that said, closing a channel when it's in the `funding_created` stage seems like a no-op so does it matter?

Reply:
> would making ready_channel idempotent be helpful in any way?

Ultimately, `ready_channel` is a call into user code, so I'm not sure we can claim anything about exactly how it behaves.

> I don't quite get the solution in the last paragraph. A ChannelError::Close return value will cause the caller to go into the closing flow. So we haven't actually prevented a close from happening with a signer that got a ready_channel, right?

I'm not *super* concerned about closing a channel post-`ready_channel` but before funding. Ultimately our counterparty can just send us an Error message and do that to us anyway. My comment about `ChannelError::Close` was more that if we return `ChannelError::Close` we're guaranteed that the `Channel` object is going to be dropped and removed ASAP, before doing anything else with it, so its arguably safe to have it be in a partially-intermediate state.

Reply:
I added the panic, which is at least useful as a warning to the operator.  However, I'm not sure this solves the issue completely.

The thing is that in general, an external signer will persist its state, so if we panic here and restart, we'll be right back where we wanted to avoid being (signer advanced state and we didn't).

I also realize that the channel is not persisted until later in the flow, so when we panic we'll forget all about it, which I suppose is fine.

The only remaining issue is that the signer may be unhappy (i.e. policy failure) that it heard about a channel that's never subsequently closed.  I suppose it can be made happy again when it sees that the funding input was re-spent for something else.

Reply:
In general, I don't think we should worry too much about correct handling of panics - if we hit them, all bets are off cause we have a severe bug somewhere and are in a state we should never get to. As for closing, one currently gaping hole in our channelmonitor logic is we have no understanding of when we are done with a monitor (or channel). Of course there's a gap between when a signer has to exist (and provide per-channel pubkeys - long before we even get to `ready_channel`) and when a monitor exists and we should consider cleaning it up.